### ISSUE TITLE
Setting stroke to transparent makes text transparent

### ISSUE THREAD

**Issue by eproxus** [excalidraw/excalidraw#5863]
When using the grid I find it useful to create transparent box shapes with text in them to align text nicely to the grid:

<img width="315" alt="Screenshot 2022-11-10 at 12 52 34" src="https://user-images.githubusercontent.com/112878/201083752-52cfb245-7d1a-4562-b501-96bbf98da191.png">

However, when selecting a box and making the stroke transparent the text becomes transparent too. One can fix this by double clicking the box (if one can find it still!) to edit the text, and select a different text color.

It would be better if setting stroke color didn't change the text color.

**Comment by webmandman** [comment]
![2022-11-11-005646_445x247_scrot](https://user-images.githubusercontent.com/991677/201303924-c8f7e7cc-0215-499f-a0d4-f05842088709.png)

Why don't I see a transparent option for the box stroke? @eproxus how do you do it?

**Comment by eproxus** [comment]
@webmandman I just guessed and it turned out you can manually enter the “hex code” `transparent` 😅

**Comment by dwelle** [comment]
@eproxus I wonder if we should fix this, as what you're undoubtedly trying to achieve here is to have a bounded text area for wrapping without the downside of using a rectangle container. Support for this will be implement in https://github.com/excalidraw/excalidraw/issues/4267.

**Comment by eproxus** [comment]
@dwelle Great to see a more fleshed out feature, although I don’t immediately see the difference between that and a normal box without a border. I’d rather have a text flow setting for the normal boxes that a completely different tool (unless I’m missing something that makes a separate tool needed). 

I only outlined one example usage where the current behavior was surprising. I’d argue that setting text color to the same as stroke color when changing it breaks the principle of least surprise no matter the intention of the user. Especially since there is a separate setting for text color.

**Comment by dwelle** [comment]
> I’d rather have a text flow setting for the normal boxes that a completely different tool (unless I’m missing something that makes a separate tool needed).

There won't be a new tool. It will work like that with the current text tool if you draw a rectangle with it, instead of clicking.

> I only outlined one example usage where the current behavior was surprising. I’d argue that setting text color to the same as stroke color when changing it breaks the principle of least surprise no matter the intention of the user. Especially since there is a separate setting for text color.

That's intentional so it's easy to quickly change the rectangle+text together, which I'd think is the more common use case.

Setting the container to transparent stroke should def be the edge case that shouldn't require this hack, but rather be solved with the text area thing outlined above.

**Comment by eproxus** [comment]
> That's intentional so it's easy to quickly change the rectangle+text together, which I'd think is the more common use case.

@dwelle I personally don’t agree with that assumption. As soon as you have selected a different text color individually I’d argue you never want this behavior.

**Comment by zsviczian** [comment]
I tend to agree with @eproxus 
I use Excalidraw every day and use containers very frequently to wrap text. Having a transparent border color for the container and a different stroke color for the text is the most frequent use case when working with sticky notes. In the Obsidian-Excalidraw implementation, I have a script that solves this for me... but it would be even nicer if I did not need a special script to achieve this behavior.

**Comment by dwelle** [comment]
>  Having a transparent border color for the container 

Right, but as I suggested above, I assume this is to handle the missing text-area/wrapping behavior of the text tool? 

>  different stroke color for the text is the most frequent use case when working with sticky notes

For sticky notes, yes. But for simple container-labeled shapes, in my experience, people want to have the same stroke for the container as for the text.

So the question is, should we have a separate tool to create sticky notes as many other apps do? At first we wanted to do so, but decided against it to keep the toolbar simple.

**Comment by eproxus** [comment]
@dwelle Let's just agree to disagree then. I know of no other tool that sets text color when stroke color is modified (but not the other way around, strangely enough). I think it's terribly inconsistent.

I personally prefer as few tools as possible, if the effect can be achieved with a flexible base tool. In this case, the tool is not so flexible and generic as I'd like.

**Comment by ngrilly** [comment]
I often have a similar need: I want to draw a rectangle with some text inside, a background color, and **no** border, which, as far as I know, is not possible today without drawing the rectangle (with a transparent stroke color) and the text separately. I'm not sure we should have a separate color for the stroke and the text. I'm wondering if a simpler solution is not to add the possibility of having a zero/none stroke width. That would also solve the use case of drawing rectangles without borders in a more "discoverable"/obvious way.

**Comment by dwelle** [comment]
> I often have a similar need: I want to draw a rectangle with some text inside, a background color, and **no** border, which, as far as I know, is not possible today without drawing the rectangle

It is, but somewhat clumsy: 1) change stroke of both 2) start editing text and change the color of the text while editing it.

I'm sure there are many other people to whom it won't be immediately obvious you can do this — which is why if we decouple the container & text colors, they won't then know how to change the text color.

---

> I'm wondering if a simpler solution is not to add the possibility of having a zero/none stroke width. That would also solve the use case of drawing rectangles without borders in a more "discoverable"/obvious way.

We would then have two different mechanisms: 1) for background, to disable it, you pick a transparent color. This also hides the background fill types. 2) for stroke, to disable it, you'd pick 0 width stroke.

Also, what if we want to later support text stroke width? We'd have the same issue as with color.

---

If we did want to decouple the container & text stroke colors, then one way to do it is show 2 color pickers for each, and perhaps separate the properties into 3 groups:

![image](https://user-images.githubusercontent.com/5153846/205972438-6040bfd6-0e23-48fa-80dc-f5e5058e37f6.png)

But that's getting a bit crazy.

And the problem is I still think the common case is people wanting to edit container & text stroke color together.

**Comment by ngrilly** [comment]
> I'm sure there are many other people to whom it won't be immediately obvious you can do this

Actually, I now remember I did this in the past, but I forgot! 😅

> if we decouple the container & text colors, they won't then know how to change the text color.

That would be great.

> We would then have two different mechanisms: 1) for background, to disable it, you pick a transparent color. This also hides the background fill types. 2) for stroke, to disable it, you'd pick 0 width stroke.
> 
> Also, what if we want to later support text stroke width? We'd have the same issue as with color.

I agree that's not a good design.

> The problem is I still think the common case is people wanting to edit container & text stroke color together.

Good point. Not sure how to solve this... Maybe the default is to keep the stroke and the text colours in sync, and "decouple" them only when the text colour is explicitly modified?... Just brainstorming here.

**Comment by andyfleming** [comment]
> If we did want to decouple the container & text stroke colors, then one way to do it is show 2 color pickers for each, and perhaps separate the properties into 3 groups:

I'm happy with this approach.

If you want the default behavior to be that it matches, you could still have that. Just have it say `match` or something in field for text color. Then it will pick up the stroke color by default, but if you change it, it will be come independent.

**Comment by indrajitbnikam** [comment]
Hi Everyone, I've created a PR that attempts to solve this issue. Can you take a look at it?

**Comment by indrajitbnikam** [comment]
Hi @maielo @dwelle , Can you have a look at my PR where I've fixed this issue?

**Comment by seanparkross** [comment]
+1

**Comment by json2d** [comment]
https://github.com/excalidraw/excalidraw/pull/6717 seems stuck on the issue of the proposed solution's alignment w/ Excalidraw's design principles.

so perhaps a fresh take w/ that in consideration:

### stated problem (refocused)
we have a shape (eg. box) w/ some text inside. we want to make the shape stroke color transparent, while keeping the text inside stroke color unchanged, but when we select the shape and set the stroke color to transparent it is applied to both.

we can followup and set the text inside stroke color directly to workaround, _but if i need to do this to many elements at once in batch then it becomes a serious time-sink_ ⏳.

### guiding principle (i presume)
if we're considering the text inside as child of the shape, this behavior is actually consistent w/ what we expect to happen when doing the same to a _grouped set of elements_ (ie. the change is applied to all elements within hierarchy)

### actual problem (reframed)
we cannot select only the parent container of shape to apply changes only to that and not to the text inside as well. 

### possible solution to actual problem
add a context menu item to **"Select parent(s)"** to reduce selection to only the root element(s).

this could then work the same for a _grouped set of elements_.

likewise we can have a context menu item to **"Select children"** to do the inverse and reduce selection to only leaf element(s).

### result

this could, in general, unlock an effective workflow for _applying hierarchically-targeted changes especially in batch_.
